### ISSUE TITLE
If a task in a block files, the always clause isn't always run

### ISSUE THREAD

**Issue by jbscare** [ansible/ansible#24301]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

block/rescue/always

##### ANSIBLE VERSION

```
ansible 2.4.0 (devel 2a7ce1059d) last updated 2017/05/04 11:57:05 (GMT -400)
  config file = /ansible/dom/ansible.cfg
  configured module search path = [u'/home/jsmift/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/jsmift/src/ansible/lib/ansible
  executable location = /home/jsmift/src/ansible/bin/ansible
  python version = 2.7.12 (default, Sep  1 2016, 22:14:00) [GCC 4.8.3 20140911 (Red Hat 4.8.3-9)]
```

##### CONFIGURATION

None that seems relevant.

##### OS / ENVIRONMENT

Amazon Linux AMI release 2016.09

##### SUMMARY

If you have a block with a task with 'run_once: True', no 'rescue' clause, but an 'always' clause, the 'always' doesn't get run if the task fails.

##### STEPS TO REPRODUCE

Playbook 1 works:

```yaml
- hosts: localhost
  gather_facts: False
  tasks:

    - block:

      - debug: msg="Help! {{ my_undefined_variable }} is not defined!"
        run_once: True

      rescue:
      - debug: msg="I'll save you!"

      always:
      - debug: msg="I'm always here for you."
```

Playbook 2 doesn't work:

```yaml
- hosts: localhost
  gather_facts: False
  tasks:

    - block:

      - debug: msg="Help! {{ my_undefined_variable }} is not defined!"
        run_once: True

      always:
      - debug: msg="I'm always here for you."
```

Playbook 3 works:

```yaml
- hosts: localhost
  gather_facts: False
  tasks:

    - block:

      - debug: msg="Help! {{ my_undefined_variable }} is not defined!"

      always:
      - debug: msg="I'm always here for you."
```

So, it works if you have 'run_once: True' and a 'rescue' clause, but not if you drop the 'rescue' clause, unless you also drop the 'run_once: True'.

##### EXPECTED RESULTS

We expected the 'always' task to always run, like this:

```
PLAY [localhost] ***************************************************************

TASK [debug] *******************************************************************
fatal: [localhost]: FAILED! => {"failed": true, "msg": "the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: 'my_undefined_variable' is undefined\n\nThe error appears to have been in '/home/jsmift/ansible/dom/playbooks/foo.yml': line 7, column 9, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n      - debug: msg=\"Help! {{ my_undefined_variable }} is not defined!\"\n        ^ here\nWe could be wrong, but this one looks like it might be an issue with\nmissing quotes.  Always quote template expression brackets when they\nstart a value. For instance:\n\n    with_items:\n      - {{ foo }}\n\nShould be written as:\n\n    with_items:\n      - \"{{ foo }}\"\n"}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "I'm always here for you."
}

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1   
```

(This is what it does if you have no 'rescue' and no 'run_once: True', like in Playbook 3.)

##### ACTUAL RESULTS

When we run Playbook 2, we get this:

```
PLAY [localhost] ***************************************************************

TASK [debug] *******************************************************************
fatal: [localhost]: FAILED! => {"failed": true, "msg": "the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: 'my_undefined_variable' is undefined\n\nThe error appears to have been in '/home/jsmift/ansible/dom/playbooks/foo.yml': line 7, column 9, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n      - debug: msg=\"Help! {{ my_undefined_variable }} is not defined!\"\n        ^ here\nWe could be wrong, but this one looks like it might be an issue with\nmissing quotes.  Always quote template expression brackets when they\nstart a value. For instance:\n\n    with_items:\n      - {{ foo }}\n\nShould be written as:\n\n    with_items:\n      - \"{{ foo }}\"\n"}

NO MORE HOSTS LEFT *************************************************************

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   
```

Note that the 'always' task wasn't run.

**Comment by jbscare** [comment]
I wonder if this is related to the bug in #19966 in some way, where a failure in a run_once task causes Ansible to bail prematurely, even if there are other hosts/tasks which could still run.

**Comment by nitzmahone** [comment]
Confirmed- this is probably something we should fix for 2.3...

**Comment by jimi-c** [comment]
Closing This Ticket
===================

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:

   * https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
   * https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
